From cd874cf8e69f1b31986fd978577994b45efa3d5c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Cl=C3=A9ment=20B=C5=93sch?= <u@pkh.me>
Date: Thu, 6 Mar 2014 18:40:03 +0100
Subject: [PATCH 12/14] http: Allow setting a Content-Type for POST requests

Signed-off-by: Luca Barbato <lu_zero@gentoo.org>
(cherry picked from commit 2572d07c1f0abd9e2bf3ed20dbe35c58f1dd6ac4)
---
 doc/protocols.texi |    3 +++
 libavformat/http.c |    6 ++++++
 2 files changed, 9 insertions(+)

diff --git a/doc/protocols.texi b/doc/protocols.texi
index 597693a..1501dab 100644
--- a/doc/protocols.texi
+++ b/doc/protocols.texi
@@ -95,6 +95,9 @@ This protocol accepts the following options:
 @item chunked_post
 If set to 1 use chunked Transfer-Encoding for posts, default is 1.
 
+@item content_type
+Set a specific content type for the POST messages.
+
 @item headers
 Set custom HTTP headers, can override built in default headers. The
 value must be a string encoding the headers.
diff --git a/libavformat/http.c b/libavformat/http.c
index b008581..7e3af5a 100644
--- a/libavformat/http.c
+++ b/libavformat/http.c
@@ -58,6 +58,7 @@ typedef struct {
     char *headers;
     char *mime_type;
     char *user_agent;
+    char *content_type;
     /* Set if the server correctly handles Connection: close and will close
      * the connection after feeding us the content. */
     int willclose;
@@ -93,6 +94,7 @@ typedef struct {
 static const AVOption options[] = {
 {"chunked_post", "use chunked transfer-encoding for posts", OFFSET(chunked_post), AV_OPT_TYPE_INT, {.i64 = 1}, 0, 1, E },
 {"headers", "set custom HTTP headers, can override built in default headers", OFFSET(headers), AV_OPT_TYPE_STRING, { 0 }, 0, 0, D|E },
+{"content_type", "set a specific content type for the POST messages", OFFSET(content_type), AV_OPT_TYPE_STRING, { 0 }, 0, 0, D|E },
 {"user_agent", "override User-Agent header", OFFSET(user_agent), AV_OPT_TYPE_STRING, {.str = DEFAULT_USER_AGENT}, 0, 0, D },
 {"user-agent", "override User-Agent header, for compatibility with ffmpeg", OFFSET(user_agent), AV_OPT_TYPE_STRING, {.str = DEFAULT_USER_AGENT}, 0, 0, D },
 {"multiple_requests", "use persistent connections", OFFSET(multiple_requests), AV_OPT_TYPE_INT, {.i64 = 0}, 0, 1, D|E },
@@ -605,6 +607,10 @@ static int http_connect(URLContext *h, const char *path, const char *local_path,
     if (!has_header(s->headers, "\r\nContent-Length: ") && s->post_data)
         len += av_strlcatf(headers + len, sizeof(headers) - len,
                            "Content-Length: %d\r\n", s->post_datalen);
+
+    if (!has_header(s->headers, "\r\nContent-Type: ") && s->content_type)
+        len += av_strlcatf(headers + len, sizeof(headers) - len,
+                           "Content-Type: %s\r\n", s->content_type);
     if (!has_header(s->headers, "\r\nIcy-MetaData: ") && s->icy) {
         len += av_strlcatf(headers + len, sizeof(headers) - len,
                            "Icy-MetaData: %d\r\n", 1);
-- 
1.7.10.4

